### ISSUE TITLE
[🐞] 关于计算loss前进行反归一化的一个疑问

### ISSUE THREAD

**Issue by 1229308805** [GestaltCogTeam/BasicTS#213]
### Is there an existing issue / discussion for this? | 是否已有关于该错误的issue或讨论？

- [X] I have searched the existing issues / discussions | 我已经搜索过已有的issues和讨论

### Is there an existing answer for this in tutorial? | 该问题是否在教程中有解答？

- [X] I have searched tutorial | 我已经搜索过tutorial

### Current Behavior | 当前行为

    在阅读tutorial时我发现，其计算流程如下：

def train_iters(data):
    data = runner.preprocessing(data)  # 归一化数据
    forward_return = runner.forward(data)  # 前向传递
    forward_return = runner.postprocessing(forward_return)  # 反归一化结果
    loss = runner.loss(forward_return)  # 计算损失
    metrics = runner.metrics(forward_return)  # 计算评估指标
    return loss

   在计算loss前进行了反归一化，此处我有一个疑问。这是约定俗成的做法吗？ 我的理解是：模型接收的是归一化后的输入，并输出归一化后的预测结果。损失函数也应该在在相同的归一化空间中计算，否则反归一化后如果不同时间序列尺度相差很大会造成影响。

### Environment | 运行环境

```Markdown
- OS:
- DEVICE:
- NVIDIA Driver:
- CUDA:
- NVIDIA GPU Memory:
- PyTorch:
```

**Comment by superarthurlx** [comment]
计算loss时是否进行反归一化是根据RESCALE这个参数决定的，在生成数据集的时候就确定了（见各个数据集的desc,json文件）。一般来说多变量数据集（Weather Electricity Traffic ETT*这些）模型接收的是归一化后的输入，并输出归一化后的预测结果，也就是RESCALE=False；时空数据集（PEMS0X这些）一般接收的是归一化后的输入，并输出反归一化后的预测结果，也就是RESCALE=True，但是也有很多工作是RESCALE=False，个人理解这个参数是跟着数据集特点决定的，目的是为了更方便地进行比较。

**Comment by 1229308805** [comment]
感谢您的回答。 刚才将RESCALE设为False在我的数据集上MAPE增大很多， 看来真的和数据集特点很相关，没有一个定论说必须或者必须不RESCALE。

**Comment by zezhishao** [comment]
> 感谢您的回答。 刚才将RESCALE设为False在我的数据集上MAPE增大很多， 看来真的和数据集特点很相关，没有一个定论说必须或者必须不RESCALE。

您好，rescale与否一般是看这个数据集大家一般怎么用，确实没有定论，符合社区习惯即可。
不过是否rescale也和metrics本身是有关系的。比如，从公式上看，mape是0值敏感的，因此归一化之后的数据一般是不能用MAPE指标的。

**Comment by 1229308805** [comment]
> > 感谢您的回答。 刚才将RESCALE设为False在我的数据集上MAPE增大很多， 看来真的和数据集特点很相关，没有一个定论说必须或者必须不RESCALE。
> 
> 您好，rescale与否一般是看这个数据集大家一般怎么用，确实没有定论，符合社区习惯即可。 不过是否rescale也和metrics本身是有关系的。比如，从公式上看，mape是0值敏感的，因此归一化之后的数据一般是不能用MAPE指标的。

原来如此，我之前忽略了这点，非常感谢您的提醒。 我现在想这样做，在训练时全程使用归一化后的数据。 在测试集上计算误差指标时使用反归一化后的数据，(目的是计算相对误差MAPE)。 这么做是否合理，我应该注意修改哪些配置参数呢

**Comment by zezhishao** [comment]
目前的代码不支持直接这么做，你需要自己修改runner，利用forward函数中的train参数，来控制postprocessing。您可以试一试，但一般来说，在什么方式上验证，就会在什么方式上计算loss。

**Comment by 1229308805** [comment]
> 目前的代码不支持直接这么做，你需要自己修改runner，利用forward函数中的train参数，来控制postprocessing。您可以试一试，但一般来说，在什么方式上验证，就会在什么方式上计算loss。

好的感谢您的解答，我已经明白了

**Comment by zezhishao** [comment]
好的，我将关闭此Issue，有任何其他关于BasicTS的问题，欢迎开一个新的issue~
